Flaky test: data race in Test/StateTransitions_SingleAddress

```
WARNING: DATA RACE
Read at 0x000000e7c030 by goroutine 224:
  google.golang.org/grpc.(*stateRecordingBalancer).HandleSubConnStateChange()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn_state_transition_test.go:468 +0x96
  google.golang.org/grpc.(*ccBalancerWrapper).watcher()
      /home/travis/gopath/src/google.golang.org/grpc/balancer_conn_wrappers.go:131 +0x608
Previous write at 0x000000e7c030 by goroutine 185:
  google.golang.org/grpc.(*stateRecordingBalancer).Build()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn_state_transition_test.go:490 +0xcb
  google.golang.org/grpc.newCCBalancerWrapper()
      /home/travis/gopath/src/google.golang.org/grpc/balancer_conn_wrappers.go:114 +0x34a
  google.golang.org/grpc.(*ClientConn).handleResolvedAddrs()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:517 +0x4f3
  google.golang.org/grpc.(*ccResolverWrapper).NewAddress()
      /home/travis/gopath/src/google.golang.org/grpc/resolver_conn_wrapper.go:116 +0x1a2
  google.golang.org/grpc/resolver/passthrough.(*passthroughResolver).start()
      /home/travis/gopath/src/google.golang.org/grpc/resolver/passthrough/passthrough.go:48 +0xc9
  google.golang.org/grpc/resolver/passthrough.(*passthroughBuilder).Build()
      /home/travis/gopath/src/google.golang.org/grpc/resolver/passthrough/passthrough.go:34 +0xca
  google.golang.org/grpc.newCCResolverWrapper()
      /home/travis/gopath/src/google.golang.org/grpc/resolver_conn_wrapper.go:89 +0x2e4
  google.golang.org/grpc.DialContext()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:291 +0xf50
  google.golang.org/grpc.testStateTransitionSingleAddress()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn_state_transition_test.go:169 +0x518
  google.golang.org/grpc.s.TestStateTransitions_SingleAddress()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn_state_transition_test.go:146 +0x66f
  google.golang.org/grpc.(*s).TestStateTransitions_SingleAddress()
      <autogenerated>:1 +0x64
  runtime.call32()
      /home/travis/.gimme/versions/go1.9.7.linux.amd64/src/runtime/asm_amd64.s:509 +0x3a
  reflect.methodValueCall()
      /home/travis/.gimme/versions/go1.9.7.linux.amd64/src/reflect/asm_amd64.s:29 +0x32
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.7.linux.amd64/src/testing/testing.go:746 +0x16c
Goroutine 224 (running) created at:
  google.golang.org/grpc.newCCBalancerWrapper()
      /home/travis/gopath/src/google.golang.org/grpc/balancer_conn_wrappers.go:113 +0x2fa
  google.golang.org/grpc.(*ClientConn).handleResolvedAddrs()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:517 +0x4f3
  google.golang.org/grpc.(*ccResolverWrapper).NewAddress()
      /home/travis/gopath/src/google.golang.org/grpc/resolver_conn_wrapper.go:116 +0x1a2
  google.golang.org/grpc/resolver/passthrough.(*passthroughResolver).start()
      /home/travis/gopath/src/google.golang.org/grpc/resolver/passthrough/passthrough.go:48 +0xc9
  google.golang.org/grpc/resolver/passthrough.(*passthroughBuilder).Build()
      /home/travis/gopath/src/google.golang.org/grpc/resolver/passthrough/passthrough.go:34 +0xca
  google.golang.org/grpc.newCCResolverWrapper()
      /home/travis/gopath/src/google.golang.org/grpc/resolver_conn_wrapper.go:89 +0x2e4
  google.golang.org/grpc.DialContext()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn.go:291 +0xf50
  google.golang.org/grpc.testStateTransitionSingleAddress()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn_state_transition_test.go:169 +0x518
  google.golang.org/grpc.s.TestStateTransitions_SingleAddress()
      /home/travis/gopath/src/google.golang.org/grpc/clientconn_state_transition_test.go:146 +0x66f
  google.golang.org/grpc.(*s).TestStateTransitions_SingleAddress()
      <autogenerated>:1 +0x64
  runtime.call32()
      /home/travis/.gimme/versions/go1.9.7.linux.amd64/src/runtime/asm_amd64.s:509 +0x3a
  reflect.methodValueCall()
      /home/travis/.gimme/versions/go1.9.7.linux.amd64/src/reflect/asm_amd64.s:29 +0x32
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.7.linux.amd64/src/testing/testing.go:746 +0x16c
Goroutine 185 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.9.7.linux.amd64/src/testing/testing.go:789 +0x568
  google.golang.org/grpc/internal/grpctest.RunSubTests()
      /home/travis/gopath/src/google.golang.org/grpc/internal/grpctest/grpctest.go:62 +0x3ef
  google.golang.org/grpc.Test()
      /home/travis/gopath/src/google.golang.org/grpc/grpc_test.go:53 +0x50
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.7.linux.amd64/src/testing/testing.go:746 +0x16c
==================
E0116 02:01:43.623412    4900 pickfirst.go:61] pickfirstBalancer: failed to NewSubConn: rpc error: code = Canceled desc = grpc: the client connection is closing
--- FAIL: Test (33.63s)
    --- FAIL: Test/StateTransitions_SingleAddress (0.12s)
    	clientconn_state_transition_test.go:145: When the server returns server preface, the client enters READY.
    	clientconn_state_transition_test.go:145: When the connection is closed, the client enters TRANSIENT FAILURE.
    	clientconn_state_transition_test.go:145: When the server sends its connection preface, but the connection dies before the client can write its
    		connection preface, the client enters TRANSIENT FAILURE.
    	clientconn_state_transition_test.go:145: When the server reads the client connection preface but does not send its connection preface, the
    		client enters TRANSIENT FAILURE.
    	testing.go:699: race detected during execution of test
	testing.go:699: race detected during execution of test
FAIL
```

--- a/clientconn_state_transition_test.go
+++ b/clientconn_state_transition_test.go
@@ -36,10 +36,10 @@ import (
 
 const stateRecordingBalancerName = "state_recoding_balancer"
 
-var testBalancer = &stateRecordingBalancer{}
+var testBalancerBuilder = newStateRecordingBalancerBuilder()
 
 func init() {
-	balancer.Register(testBalancer)
+	balancer.Register(testBalancerBuilder)
 }
 
 // These tests use a pipeListener. This listener is similar to net.Listener
@@ -148,9 +148,6 @@ client enters TRANSIENT FAILURE.`,
 }
 
 func testStateTransitionSingleAddress(t *testing.T, want []connectivity.State, server func(net.Listener) net.Conn) {
-	stateNotifications := make(chan connectivity.State, len(want))
-	testBalancer.ResetNotifier(stateNotifications)
-
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
 
@@ -172,6 +169,8 @@ func testStateTransitionSingleAddress(t *testing.T, want []connectivity.State, s
 		t.Fatal(err)
 	}
 	defer client.Close()
+
+	stateNotifications := testBalancerBuilder.nextStateNotifier()
 
 	timeout := time.After(5 * time.Second)
 
@@ -204,9 +203,6 @@ func (s) TestStateTransitions_ReadyToTransientFailure(t *testing.T) {
 		connectivity.TransientFailure,
 		connectivity.Connecting,
 	}
-
-	stateNotifications := make(chan connectivity.State, len(want))
-	testBalancer.ResetNotifier(stateNotifications)
 
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
@@ -247,6 +243,8 @@ func (s) TestStateTransitions_ReadyToTransientFailure(t *testing.T) {
 	}
 	defer client.Close()
 
+	stateNotifications := testBalancerBuilder.nextStateNotifier()
+
 	timeout := time.After(5 * time.Second)
 
 	for i := 0; i < len(want); i++ {
@@ -271,9 +269,6 @@ func (s) TestStateTransitions_TriesAllAddrsBeforeTransientFailure(t *testing.T) 
 		connectivity.Connecting,
 		connectivity.Ready,
 	}
-
-	stateNotifications := make(chan connectivity.State, len(want))
-	testBalancer.ResetNotifier(stateNotifications)
 
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
@@ -334,6 +329,8 @@ func (s) TestStateTransitions_TriesAllAddrsBeforeTransientFailure(t *testing.T) 
 	}
 	defer client.Close()
 
+	stateNotifications := testBalancerBuilder.nextStateNotifier()
+
 	timeout := time.After(5 * time.Second)
 
 	for i := 0; i < len(want); i++ {
@@ -368,9 +365,6 @@ func (s) TestStateTransitions_MultipleAddrsEntersReady(t *testing.T) {
 		connectivity.TransientFailure,
 		connectivity.Connecting,
 	}
-
-	stateNotifications := make(chan connectivity.State, len(want))
-	testBalancer.ResetNotifier(stateNotifications)
 
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
@@ -431,6 +425,8 @@ func (s) TestStateTransitions_MultipleAddrsEntersReady(t *testing.T) {
 	}
 	defer client.Close()
 
+	stateNotifications := testBalancerBuilder.nextStateNotifier()
+
 	timeout := time.After(2 * time.Second)
 
 	for i := 0; i < len(want); i++ {
@@ -454,42 +450,53 @@ func (s) TestStateTransitions_MultipleAddrsEntersReady(t *testing.T) {
 }
 
 type stateRecordingBalancer struct {
-	mu       sync.Mutex
 	notifier chan<- connectivity.State
-
 	balancer.Balancer
 }
 
 func (b *stateRecordingBalancer) HandleSubConnStateChange(sc balancer.SubConn, s connectivity.State) {
-	b.mu.Lock()
 	b.notifier <- s
-	b.mu.Unlock()
-
 	b.Balancer.HandleSubConnStateChange(sc, s)
 }
 
 func (b *stateRecordingBalancer) ResetNotifier(r chan<- connectivity.State) {
-	b.mu.Lock()
-	defer b.mu.Unlock()
 	b.notifier = r
 }
 
 func (b *stateRecordingBalancer) Close() {
-	b.mu.Lock()
-	u := b.Balancer
-	b.mu.Unlock()
-	u.Close()
+	b.Balancer.Close()
 }
 
-func (b *stateRecordingBalancer) Name() string {
+type stateRecordingBalancerBuilder struct {
+	mu       sync.Mutex
+	notifier chan connectivity.State // The notifier used in the last Balancer.
+}
+
+func newStateRecordingBalancerBuilder() *stateRecordingBalancerBuilder {
+	return &stateRecordingBalancerBuilder{}
+}
+
+func (b *stateRecordingBalancerBuilder) Name() string {
 	return stateRecordingBalancerName
 }
 
-func (b *stateRecordingBalancer) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+func (b *stateRecordingBalancerBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	stateNotifications := make(chan connectivity.State, 10)
 	b.mu.Lock()
-	b.Balancer = balancer.Get(PickFirstBalancerName).Build(cc, opts)
+	b.notifier = stateNotifications
 	b.mu.Unlock()
-	return b
+	return &stateRecordingBalancer{
+		notifier: stateNotifications,
+		Balancer: balancer.Get(PickFirstBalancerName).Build(cc, opts),
+	}
+}
+
+func (b *stateRecordingBalancerBuilder) nextStateNotifier() <-chan connectivity.State {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	ret := b.notifier
+	b.notifier = nil
+	return ret
 }
 
 type noBackoff struct{}

--- a/clientconn_test.go
+++ b/clientconn_test.go
@@ -1003,9 +1003,6 @@ func (s) TestUpdateAddresses_RetryFromFirstAddr(t *testing.T) {
 	server2ContactedSecondTime := make(chan struct{})
 	server3Contacted := make(chan struct{})
 
-	stateNotifications := make(chan connectivity.State, 100)
-	testBalancer.ResetNotifier(stateNotifications)
-
 	// Launch server 1.
 	go func() {
 		// First, let's allow the initial connection to go READY. We need to do
@@ -1024,6 +1021,10 @@ func (s) TestUpdateAddresses_RetryFromFirstAddr(t *testing.T) {
 			return
 		}
 
+		// nextStateNotifier() is updated after balancerBuilder.Build(), which is
+		// called by grpc.Dial. It's safe to do it here because lis1.Accept blocks
+		// until balancer is built to process the addresses.
+		stateNotifications := testBalancerBuilder.nextStateNotifier()
 		// Wait for the transport to become ready.
 		for s := range stateNotifications {
 			if s == connectivity.Ready {

